openlibrary/accounts/model.py

```diff
@@ -597,42 +604,90 @@ def audit_accounts(email, password, test=False):
-    elif ia_login == "ok":
+    if ia_login == "ok":
         ia_account = InternetArchiveAccount.get(email=email, test=test)
 
-        audit['authenticated'] = 'ia'
-        audit['has_ia'] = email
+        audit['authenticated'] = 'ia'  # XXX authenticated_with
+        audit['has_ia'] = email  # XXX change to ia_email
 
-        # Get the OL account which links to this IA account, if
-        # one exists (i.e. this IA account could be linked to an
-        # OL account with a different email)
+        # Get the OL account which links to this IA account
         ol_account = OpenLibraryAccount.get(link=ia_account.itemname, test=test)
         link = ol_account.itemname if ol_account else None
 
-        # If no linked account was found but there's an ol_account
-        # having the same email as this IA account, mark them to
-        # be linked
+        # The fact that there is no link implies no Open Library
+        # account exists containing a link to this Internet Archive
+        # account...
         if not link:
+            # then check if there's an Open Library account which shares
+            # the same email as this IA account.
             ol_account = OpenLibraryAccount.get(email=email, test=test)
-            link = ia_account.itemname if ol_account else None
 
-        # So long as there's either a linked OL account, or an OL
+            # If an Open Library account with a matching email account exist...
+            if ol_account:
+                # Check whether it is linked already, i.e. has an itemname
+                # set. We already determined that no OL account is
+                # linked to our IA account. Therefore this Open
+                # Library account having the same email as our IA
+                # account must have been linked to a different
+                # Internet Archive account.
+                if ol_account.itemname:
+                    return {'error': 'wrong_ia_account'}
+
+        # At this point, it must either be the case that (a)
+        # `ol_account` already links to our IA account (in which case
+        # `link` has a correct value), (b) that an unlinked
+        # `ol_account` shares the same email as our IA account and
+        # thus can and should be safely linked to our IA account, or
+        # (c) no `ol_account` which is linked or can be linked has
+        # been found and therefore, asuming
+        # lending.config_ia_auth_only is enabled, we need to create
+        # and link it.
+        if not ol_account:
+            if lending.config_ia_auth_only:
+                try:
+                    ol_account = OpenLibraryAccount.create(
+                        ia_account.itemname, email, password,
+                        displayname=ia_account.screenname,
+                        verified=True, test=test)
+                except ValueError as e:
+                    return {'error': 'max_retries_exceeded', 'msg': str(e)}
+
+                if not ol_account.itemname:
```

Comment:
You're right, there shouldn't be a case possible, I added this because a function had been removed between branches causing it to appear as if the account did not have an itemname set after login (when really the function to fetch the itemname was indeed the problem)